### PR TITLE
Add contains_url filter option and /messages filter

### PR DESCRIPTION
**Pull request by erikjohnston**
I'm not really sure the best way of spec'ing the filter param on `/messages`, suggestions welcome.

### CODE REVIEW

**Comment by richvdh** [comment]
Supercedes https://docs.google.com/document/d/1J-Bd9QM_h6pH3xLidsIPsyNRQZGc3VfeKdy6O8bQLHI/edit, presumably?

**Comment by erikjohnston** [comment]
yup

File: api/client-server/definitions/room_event_filter.yaml [hunk 1]
```diff
@@ -28,4 +28,8 @@ properties:
     items:
       type: string
     type: array
+  contains_url:
+    type: boolean
+    description: If specified will filter events based on if they have a url key in their
```
Comment (richvdh) [inline api/client-server/definitions/room_event_filter.yaml]: "If ``true``, includes only events with a ``url`` key in their ``content``. If ``false``, excludes those events"

maybe?

File: api/client-server/message_pagination.yaml [hunk 2]
```diff
@@ -75,6 +75,11 @@ paths:
           description: |-
             The maximum number of events to return. Default: 10.
           x-example: "3"
+        - in: query
+          type: string
+          name: filter
+          description: |-
+            A url encoded JSON RoomEventFilter to filter returned events with.
```
Comment (richvdh) [inline api/client-server/message_pagination.yaml]: well, of course it's url-encoded. it's a query-param. I don't think there is a need to specify this. Unless you mean it's double-encoded? 

I would just say "A JSON-encoded RoomEventFilter to filter returned events with."

But the main thing it needs is an example.

File: changelogs/client_server.rst [hunk 3]
```diff
@@ -35,6 +35,10 @@
   - Add top-level ``account_data`` key to the responses to ``GET /sync`` and
     ``GET /initialSync``
     (`#380 <https://github.com/matrix-org/matrix-doc/pull/380>`_).
+  - Add ``contains_url`` to ``RoomEventFilter``.
```
Comment (richvdh) [inline changelogs/client_server.rst]: "Add ``contains_url`` _option_" ?

File: api/client-server/definitions/room_event_filter.yaml (hunk 1 above)
Comment (erikjohnston) [inline api/client-server/definitions/room_event_filter.yaml]: Done

File: api/client-server/message_pagination.yaml (hunk 2 above)
Comment (erikjohnston) [inline api/client-server/message_pagination.yaml]: Done

File: changelogs/client_server.rst (hunk 3 above)
Comment (erikjohnston) [inline changelogs/client_server.rst]: Done

**Pull request merged**